ch_pipeline/hfb/flagging.py

```diff
@@ -0,0 +1,145 @@
+            frequencies, subfrequencies and time samples.
+        """
+
+        # Extract data and weight arrays, averaging over beams
+        data = np.mean(stream.hfb[:], axis=2)
+        weight = np.mean(stream.weight[:], axis=2)
+
+        # Number of samples per data point in the HFB data:
+        # N_S = Delta nu * Delta t, where Delta nu is the frequency resolution
+        # (390.625 kHz / 128) and Delta t the integration time (10 s)
+        # NOTE: Should this be extracted from the container attributes? In that
+        # case this information needs to be added to the attributes upstream.
+        N_SAMP = 390625.0 / 128.0 * 10.0
+
+        # Ideal radiometer equation
+        radiometer = data**2 / N_SAMP
+
+        # Radiometer noise test: the sensitivity metric would be unity for
+        # an ideal radiometer, it would be higher for data with RFI
+        sensitivity_metric = 2.0 / (radiometer * weight)
+
+        # Boolean mask idicating data that are contaminated by RFI
+        mask = sensitivity_metric > self.threshold
+
+        # Create container to hold output
+        out = HFBRFIMask(axes_from=stream, attrs_from=stream)
+
+        if self.keep_sens:
+            out.add_dataset("sens")
+            out.sens[:] = sensitivity_metric
+
+        # Save mask to output container
+        out.mask[:] = mask
+
+        # Add beam selection to RFI mask attributes
+        out.attrs["beam"] = stream._data["index_map"]["beam"][:]
+
+        # Return output container
+        return out
+
+
+class ApplyHFBMask(task.SingleTask):
+    """Apply a mask to an HFB stream.
+
+    Attributes
+    ----------
+    zero_data : bool, optional
+        Zero the data in addition to modifying the weights. Default is True.
+    """
+
+    zero_data = config.Property(proptype=bool, default=True)
+
+    def process(self, stream, mask):
+        """Set weights to zero for flagged data.
+
+        Parameters
+        ----------
+        stream : containers.HFBData
+            Container with HFB data and weights.
+
+        mask : containers.HFBRFIMask
+            Boolean mask indicating the samples that are contaminated by RFI.
+
+        Returns
+        -------
+        stream : containers.HFBData
+            Container with HFB data and weights, with weights of flagged data
+            set to zero."""
+
+        # flag = mask[:, :, np.newaxis, :]
+
+        # Create a slice that will expand the mask to
+        # the same dimensions as the weight array
+        waxis = stream.weight.attrs["axis"]
+        slc = [slice(None)] * len(waxis)
+        for ww, name in enumerate(waxis):
+            if name not in mask.mask.attrs["axis"]:
+                slc[ww] = None
+
+        # Extract mask
+        flag = mask.mask[:].astype(stream.weight.dtype)
+
+        # Convert mask from caput.mpiarray.MPIArray to regular numpy.ndarray
+        flag = np.array(flag)
+
+        # Expand mask to same dimension as weight array
+        flag = flag[tuple(slc)]
```

Comment:
I don't think all of this is strictly necessary, i.e. I don't think you need to upcast, and I also don't think you need to transform the array into a regular numpy array (and if you do, you should just do `flag.local_array`).

You might need to be a little careful with the MPIArray vs numpy array aspects when printing out the fraction flagged.

Reply:
Added a `TODO` comment in the code.